### PR TITLE
Relaod using magazines [WIP] [CR]

### PR DESCRIPTION
**Pull request by mugling**
The aims of this PR is to define the minimal specification for a magazine and allow it to be reloaded and unloaded. A second PR can then extend the reloading code so that guns have either integral (default) or detachable magazines. Subsequent PR's can then start to implement specific magazines.

In this implementation magazines store one or more ammo items within their `contents` vector. At present these must be of the same type (as returned by `item::ammo_current`) so contained ammo should only differ in terms of damage or other such properties. `item::get_curammo_id` is deprecated in favour of `item::ammo_current` which is extended to correctly handle magazines with contained ammo.

A later refactor of `game::unload` can probably combine unloading of magazines and containers but such refinements are possibly best left until a minimal working implementation can be committed.

TODO:
- [X] Magazines can be repeatedly loaded with equivalent ammo until full
- [X] Magazines when unloaded return appropriately stacked ammo
- [X] Loaded magazines increase in weight proportional to contained ammo
- [X] Guns presently ignore ammo contained within magazines
- [ ] Info for magazines is correctly displayed

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -11442,6 +11442,22 @@ void game::unload( item &it )
         return;
     }
 
+    // Handle magazines as a special case
+    if( target->is_magazine() ) {
+        // Remove all contained ammo consuming half as much time as required to load the magazine
+        target->contents.erase( std::remove_if( target->contents.begin(), target->contents.end(), [&]( item& e ) {
```
Comment (mugling) [inline src/game.cpp]: Any reason we can't empty the magazine in LIFO order via contents.rbegin() and contents.rend() here?

Comment (Coolthulhu) [inline src/game.cpp]: Erase-Remove is cleaner than manual iteration.
Do you need manual iteration here for any reason?

File: data/json/items/magazines.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[
+    {   "type" : "MAGAZINE",
+        "id" : "stanag30",
+        "price" : 8000,
+        "name" : "STANAG magazine",
```
Comment (Coolthulhu) [inline data/json/items/magazines.json]: Mentioning the caliber would be useful.

**Comment by SeanMirrsen** [comment]
Must.... resist... pointing... out... typo....

**Comment by chaosvolt** [comment]
Ah, will be interesting to see.

Also, regarding your second PR, I'm hoping the idea of "old method AND/OR new method" has been discussed. This would allow for things like revolver speedloaders and stripper clips for rifles.

File: src/game.cpp (hunk 1 above)
Comment (mugling) [inline src/game.cpp]: I'm uncertain if the erase-remove idiom is safe to use with reverse iterators

**Comment by mugling** [comment]
@SeanMirrsen indeed

@chaosvolt yes, the implementation will explicitly support that (`use_ammo` and `use_mags` both set to true)

**Comment by chaosvolt** [comment]
> Must.... resist... pointing... out... typo....

Doh. I only just now noticed the title typo. XD

**Comment by SeanMirrsen** [comment]
Can magazines be made non-reloadable and non-unloadable? Either at all or unless a special action/item is used?
This is mostly in regards to batteries, and changing cash cards to a similar system.

**Comment by mugling** [comment]
> Can magazines be made non-reloadable and non-unloadable? Either at all or unless a special action/item is used?

Yes, this is already supported in `item::reload`

File: src/game.cpp (hunk 1 above)
Comment (Coolthulhu) [inline src/game.cpp]: Why do you want reverse iterators here?
Do you want to cancel reloading at some point?

**Comment by SeanMirrsen** [comment]
Further comment. Say you have a watergun. Yes, useless ingame (unless it's an acid watergun). Just in principle. It has a "magazine" that is a plastic bottle of water. Would you be able to use a plastic bottle as a magazine, and water as ammo, without jumping through hoops?

What I mean is, is the system interchangeable with the current "item with charges" system? If I wanted to make a meat sandwich cannon, would I be able to plug three charges of meat sandwich into a paper wrapper magazine and "fire" it?

**Comment by chaosvolt** [comment]
What next, cheese cannonballs?

Though this sounds like the sort of suggestion that implies the future option to expanding this planned system to allow other options for "things inside things" until eventually we have that backpack idea using similar mechanisms (though ideally streamlined in whatever way makes it the least annoying to use).

**Comment by SeanMirrsen** [comment]
Baby steps. :)
Right now this is only for things that are defined as "ammo", as far as I can tell. It won't allow to load different ammo stacks in. But, in a roundabout way, yes. Things inside things.

**Comment by mugling** [comment]
@SeanMirrsen No, magazines are distinct from containers. What you are suggesting would require an entirely new implementation of containers with magazines implemented as a subtype. It's possible but more work than anyone has presently shown interest in completing.

Incidentally some work on nested containers was begun in #14600 however we still lack consensus as to a sane UI for implementing this.

The aim of this PR is to determine the minimum specification for a magazine and check reloading works without any bugs.

**Comment by SeanMirrsen** [comment]
>however we still lack consensus as to a sane UI for implementing this

As far as I can tell there is no *sane* UI. Keeping it simple and showing container contents is nigh mutually exclusive. The only good option I can think of is a toggle. Default view as current, just an item list. Advanced view as a scrollable tree of containers using the ASCII window-edge symbols to show what is where.

And sure, I know it's outside the scope of the PR. Just wondering how far away it is. :)

File: src/game.cpp (hunk 1 above)
Comment (mugling) [inline src/game.cpp]: Not to cancel but because `add_or_drop_with_msg` prompts for empty containers where necessary and could fail if earlier unloaded ammo has consumed them all.

It's a trivial detail but one I'd like to get correct now if at all possible.

**Comment by mugling** [comment]
> Just wondering how far away it is. :)

From what I understand the UI is the major stumbling block, see http://smf.cataclysmdda.com/index.php?topic=10444.0

**Comment by SeanMirrsen** [comment]
As far as I can see the major stumbling block is someone who'd take up the flag and carry it. :P 
There would need to be a project first, in order for people to try and hammer out the details.

**Comment by chaosvolt** [comment]
Right. Jury-rigging this idea to allow for containers might be awkward. That said, lessons learned during the work on this may still prove useful for future inventory shenanigans. XP

File: src/item.cpp [hunk 3]
```diff
@@ -3965,6 +3984,19 @@ bool item::ammo_consume( int qty ) {
     return false;
 }
 
+itype_id item::ammo_current() const
+{
+    if( is_ammo() ) {
+        return typeId();
```
Comment (BevapDin) [inline src/item.cpp]: Documentation says "Specific ammo type, returns "null" if item is neither ammo nor loaded with any", but this returns an empty string instead.

File: src/item.cpp [hunk 4]
```diff
@@ -3965,6 +3984,19 @@ bool item::ammo_consume( int qty ) {
...
+    }
+
+    if( is_magazine() ) {
+        return contents.size() ? contents[0].type->id : "null";
```
Comment (BevapDin) [inline src/item.cpp]: What has size to do here, didn't you mean `!empty()`?

File: src/item.cpp (hunk 3 above)
Comment (mugling) [inline src/item.cpp]: Noted. Incidentally we use `"NULL"` and `"null"` throughout the codebase. Would there be any objection to a separate PR that converts everything to the same and if so which is the preferred form?

File: src/item.cpp (hunk 4 above)
Comment (mugling) [inline src/item.cpp]: Logically equivalent however I agree `!contents.empty()` is more semantically correct

File: src/item.cpp (hunk 3 above)
Comment (mugling) [inline src/item.cpp]: Does 6fb1b9c resolve this? Is `type->id` ever an empty string?

Comment (BevapDin) [inline src/item.cpp]: > Incidentally we use  "NULL"  and  "null"  throughout the codebase.

That's why I came up with `string_id<T>::NULL_ID` (see string_id.h). It is, however, of the type `string_id<T>`, not just a plain `std::string`. Using it avoids having to remember what the null id actually is.

For example: it's "mon_null" for monster type, but that string appears exactly once as string literal in the source, all other places use `NULL_ID`.

> Would there be any objection to a separate PR that converts everything to the same and if so which is the preferred form?

I would avoid that, to much work. You'll have to change various JSON files or translate the old null-ids to new null-ids.

> Is  type->id  ever an empty string?

`type->id` is always a valid item type id, so yes, this should work fine.

Comment (mugling) [inline src/item.cpp]: > That's why I came up with string_id<T>::NULL_ID

I did come across this but wasn't sure as to the usage. This can be used as a replacement for checking `id` == `null`?

Comment (BevapDin) [inline src/item.cpp]: Basically, once you implemented `NULL_ID` like this (see trap.cpp for an actual example):
```C++
template<>
const string_id<Foo> string_id<Foo>::NULL_ID( "foo_null_id" );
```

You can now use `string_id<Foo>::is_null` and even the explicit `operator bool`:

```C++
string_id<Foo> id = ...;
if( id ) {
    // id is not null
} else {
    // id is equal to the null-id
}

// Or for function returning such:
string_id<Foo> get_foo_id() {
    if( some_condition ) {
        return valid_foo_id_from_somewhere();
    } else {
        return string_id<Foo>::NULL_ID;
    }
}
```

The only place you may need to document the actual value of the null id is on the JSON documentation when it's actually read from there.
